### PR TITLE
fix(core): add support for structs in constructor args

### PR DESCRIPTION
**Pull request by sam-goldman**

### CODE REVIEW

File: packages/core/src/utils.ts [hunk 1]
```diff
@@ -831,43 +831,35 @@ export const getContractAddress = (
...
-  )
+  const constructorArgValues = getConstructorArgs(constructorArgs, abi)
+
+  const iface = new ethers.utils.Interface(abi)
 
   const creationCodeWithConstructorArgs = bytecode.concat(
-    remove0x(
-      utils.defaultAbiCoder.encode(constructorArgTypes, constructorArgValues)
-    )
+    remove0x(iface.encodeDeploy(constructorArgValues))
```
Comment (sam-goldman) [inline packages/core/src/utils.ts:862]: I replaced `defaultAbiCoder.encode` with `iface.encodeDeploy` because `defaultAbiCoder.encode` would silently fail if the `constructorArgTypes` weren't defined correctly

File: packages/core/src/config/parse.ts [hunk 2]
```diff
@@ -1958,7 +1958,10 @@ export const assertValidContracts = (
         }
       }
 
-      if (!contractConfig.unsafeAllowEmptyPush) {
+      if (
+        !contractConfig.unsafeAllowEmptyPush &&
+        contractConfig.kind !== 'no-proxy'
```
Comment (RPate97) [inline packages/core/src/config/parse.ts:1963]: What's the purpose of this change?

Comment (sam-goldman) [reply packages/core/src/config/parse.ts:1994]: It's not related to the main purpose of this PR. It just skips the `unsafeAllowEmptyPush` check for non-proxied contracts. This is because `push()` can only be dangerous in an upgradeable contract.

File: packages/core/src/config/parse.ts [hunk 3]
```diff
@@ -1262,33 +1303,82 @@ export const parseStructConstructorArg = (
...
+      'warning',
+      `Detected array(s) in struct constructor argument. Struct members that are arrays will be ignored by the Solidity compiler and cannot be set in the constructor.`,
+      [
+        'The following struct members will be ignored: ',
+        ...arrayMembers.map((el) => `${el.name}: ${el.paramType.type}`),
+      ],
+      cre.silent,
+      cre.stream
+    )
+  }
```
Comment (sam-goldman) [inline packages/core/src/config/parse.ts:1328]: I actually don't think this is necessary. The Solidity compiler doesn't ignore these values. I ran into this same issue a while ago, and it turns out that you just need to explicitly define a getter function for them in Solidity. See [here](https://github.com/chugsplash/chugsplash/blob/develop/packages/contracts/contracts/ChugSplashManager.sol#L740-L754) for an explanation (esp the linked github issue)

Comment (RPate97) [reply packages/core/src/config/parse.ts:1328]: Wow, that's pretty wonky

Comment (sam-goldman) [reply packages/core/src/config/parse.ts:1328]: Lol yep, welcome to Solidity
